exo-setup/src/docker-setup.ls

```diff
@@ -1,67 +1,92 @@
 }
 
 
-class DockerSetup extends EventEmitter
-
-  ({@role, @logger, @config}) ->
-    @service-config = if @config.root then yaml.safe-load fs.read-file-sync(path.join(@config.root, 'service.yml'), 'utf8')
-
-
-  start: (done) ~>
-    | !@service-config        =>  return @_setup-external-service done
-    | !@_docker-file-exists!  =>  cp path.join(templates-path, 'docker', 'Dockerfile'), path.join(@config.root, 'Dockerfile')
-
-    @logger.log {@role, text: "preparing Docker image"}
-    @_build-docker-image done
-
-
-  _build-docker-image: (done) ~>
-    new ObservableProcess(call-args(DockerHelper.get-build-command author: @service-config.author, name: dashify(@service-config.type))
-                          cwd: @config.root
-                          stdout: {@write}
-                          stderr: {@write})
-      ..on 'ended', (exit-code, killed) ~>
-        | exit-code is 0  =>  @logger.log {@role, text: "Docker setup finished"}
-        | otherwise       =>
-          @logger.log {@role, text: "Docker setup failed"}
-          process.exit exit-code
-        done!
-
-
-  _docker-file-exists: ~>
-    try
-      fs.exists-sync path.join(@config.root, 'Dockerfile')
-    catch
-      no
-
-
-  _setup-external-service: (done) ~>
-    throw new Error red "No location or docker-image specified" unless @config.docker-image
-    image = @config.docker-image |> (.split '/')
-    new ObservableProcess((DockerHelper.get-pull-command author: image[0], name: image[1]),
-                          stdout: {@write}
-                          stderr: {@write})
-      ..on 'ended', (exit-code) ~>
-        | exit-code isnt 0  =>
-          @logger.log {@role, text: 'Docker setup failed'}
-          process.exit exit-code
-        | _                 =>  done!
-
-
-  write: (text) ~>
-    @logger.log {@role, text, trim: yes}
-
+# Renders docker-compose.yml file with service configuration
+class DockerSetup
+
+  ({@app-name, @role, @logger, @service-location, @docker-image}) ->
+    @service-config = yaml.safe-load fs.read-file-sync(path.join(process.cwd!, @service-location, 'service.yml'), 'utf8') if @service-location
+
+
+  get-service-docker-config: ~> 
+    | @service-config => @_get-service-docker-config!
+    | otherwise       => @_get-external-service-docker-config!
+
+
+  # builds the Docker config for a service and its dependencies
+  _get-service-docker-config: ->
+    docker-config = {}
+    docker-config[@role] = Obj.compact do
+      build: path.join '..', @service-location 
+      container_name: @role
+      command: @service-config.startup.command
+      ports: @service-config.docker?.ports or undefined
+      links: @_get-docker-links!
+      environment: @_get-docker-env-vars! 
+      depends_on: @_get-service-dependencies!
+    for dependency, dependency-config of @service-config.dependencies
+      docker-config[dependency + dependency-config.dev.version] = @_get-service-dependency-docker-config dependency, dependency-config.dev
```

Comment:
I think `+` makes sense here, since we aren't using literal string characters it doesn't seem necessary to enter into a string just to escape it again.